ingest-router/src/project_config/splitter.rs

```diff
@@ -0,0 +1,241 @@
+//! Key routing logic for distributing requests across upstream cells.
+//!
+//! Uses the locator service to route public keys to their owning cells.
+
+use crate::locale::Cells;
+use locator::client::{ClientError, Locator};
+use locator::locator::LocatorError;
+use std::collections::HashMap;
+
+#[allow(dead_code)]
+pub type PublicKey = String;
+#[allow(dead_code)]
+pub type CellId = String;
+
+/// Routes public keys to their owning cells using the locator service.
+#[allow(dead_code)]
+pub struct PublicKeySplitter {
+    locator: Locator,
+}
+
+#[allow(dead_code)]
+impl PublicKeySplitter {
+    pub fn new(locator: Locator) -> Self {
+        Self { locator }
+    }
+
+    /// Splits public keys to their owning cells using locator service lookups.
+    ///
+    /// For each public key, queries the locator to determine which cell owns it.
+    /// Returns a map of cell IDs to their public keys, plus a list of keys that
+    /// couldn't be routed (no cell found, locator errors, or unconfigured cells).
+    pub async fn split(
+        &self,
+        public_keys: &[PublicKey],
```

Comment:
how about passing the owned public keys in here, then you can skip all the clones inline?

```suggestion
        public_keys: Vec<PublicKey>,
```